oak/server/rust/oak_runtime/src/time/mod.rs

```diff
@@ -0,0 +1,383 @@
+//
+// Copyright 2020 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use futures::future::join_all;
+use log::{debug, info, warn};
+use roughenough::{
+    client::{create_nonce, make_request, ParsedResponse, ResponseHandler},
+    RtMessage,
+};
+use std::{
+    net::{SocketAddr, ToSocketAddrs as _},
+    time::Duration,
+};
+use tokio::{net::UdpSocket, runtime::Runtime, time::timeout};
+
+#[cfg(test)]
+mod tests;
+
+pub const DEFAULT_MIN_OVERLAPPING_INTERVALS: usize = 3;
+pub const DEFAULT_MAX_RADIUS_MICROSECONDS: u32 = 60000000;
+pub const DEFAULT_TIMEOUT_SECONDS: u64 = 3;
+pub const DEFAULT_SERVER_RETRIES: usize = 3;
+const MAX_RESPONSE_SIZE: usize = 1024;
+
+/// Specifies the details of a Roughtime server.
+///
+/// Only UDP is supported as a protocol and ED25519 as a public key type.
+pub struct RoughtimeServerSpec {
+    /// The name of the Roughtime server.
+    name: String,
+    /// The address of the Roughtime server.
+    host: String,
+    /// The port on which the Roughtime server is listening.
+    port: u16,
+    /// The base64-encoded public key for the Roughtime server.
+    public_key_base64: String,
+}
+
+/// A client for requesting Roughtime from multiple servers.
+///
+/// Time is given as microseconds since the UNIX epoch (00:00:00 UTC on 1 January 1970).
+/// Leap seconds are linearly smeared over a 24-hour period. That is, the smear extends from
+/// UTC noon to noon over 86,401 or 86,399 SI seconds, and all the smeared seconds are the same
+/// length.
+pub struct RoughtimeClient {
```

Comment:
I wonder if anything about this should be added to the docs. But perhaps that is something that can be done in a separate PR.

Reply:
This version will only be used by the runtime. App developers will interact with the Roughtime pseudo-node, for which there is already some documentation. I will check that the documentation is current as part of the next PR that will implement the Rust version of the Roughtime pseudo-node.